### ISSUE TITLE
Can't initialize a distinct type with a generic

### ISSUE THREAD

**Issue by dawkot** [nim-lang/Nim#6248]
```
type
    A[T] = object
        x: T

    B[T] = distinct A[T]

var x = B[int]() #error: object constructor needs an object type
```

**Comment by Araq** [comment]
Well yes, the object construction syntax is for objects, a distinct object type is not an object type.

**Comment by andreaferretti** [comment]
Wow, this was not clear at all! Makes sense, but I think this should be noted in the manual when talking about distinct types

**Comment by dom96** [comment]
I don't think you can borrow `.` like that. The `.` isn't overloaded for that type so there is nothing to borrow, the built-in `.` isn't borrowable.

**Comment by zah** [comment]
Why isn't this considered an issue? Most people will be surprised that the following code doesn't work:

```nim
type
  Foo = object
    x: int

  Bar = distinct Foo

var a = Foo(x: 10)
var b = Bar(x: 10)
```

... and I don't see many downsides to allowing it. One issue that I stumbled upon in our own codebase was that I wasn't able to use a distinct type with a generic proc expecting to construct the passed in type. There is no simple solution in this situation except abandoning the construction syntax for either field-by-field initialization or a separate `init` proc. Over time, silly things like this will lead Scott Meyers to write a book called "Effective Nim", where he will argue that you should avoid using the object construction syntax, because it may lead to problems down the road. This, after all the fight from us to skip the chapter on getters and setters :)

**Comment by Araq** [comment]
What if I change the implementation from ``distinct Foo`` to ``distinct int``? Then the construction syntax stops working. I have no idea why it should be allowed, a ``distinct`` type is for data abstraction and information hiding. And why would Scott Meyers not understand this? This design decision is easy enough to understand and to justify.

**Comment by zah** [comment]
The problem is that I cannot create a distinct type that allows the constructor syntax if that's what I want. Perhaps we can allow borrowing the constructor just like we allow borrowing for `.`?

Consider types such as `AbsoluteDir` and `RelativeFile`. Part of their contract is that you can construct them from a string value, but that's fine - it's convenient and you believe that it will remain true in the future.

**Comment by Araq** [comment]
> Perhaps we can allow borrowing the constructor just like we allow borrowing for .?

That would be fine. Yet another feature though.

**Comment by haoyu234** [comment]
I also ran into this problem where I couldn't construct a distinct object through the constructor syntax, which confused me.

```nim
type
  InstruQueueNode* = object
    previous*: ptr InstruQueueNode
    next*: ptr InstruQueueNode

  InstruQueue* = distinct InstruQueueNode
```

```shell
/tests/tqueue.nim(23, 1) template/generic instantiation of `test` from here
/tests/tqueue.nim(24, 22) Error: object constructor needs an object type [object declared in /src/instru/queue.nim(4, 3)]
tools.nim(36)            doCmd
```

<del>I think I can solve this problem by copying and pasting the fields.</del>

edit:
I solved this problem by using the `InstruQueueNode` object as a field of the `InstruQueue` object.
```nim
  InstruQueue* = object
    node*: InstruQueueNode
```
